pkg/dedup/iter.go

```diff
@@ -380,32 +380,32 @@ type dedupSeriesIterator struct {
 
 	// TODO(bwplotka): Don't base on LastT, but on detected scrape interval. This will allow us to be more
 	// responsive to gaps: https://github.com/thanos-io/thanos/issues/981, let's do it in next PR.
-	lastT int64
-	lastV float64
+	lastT    int64
+	lastIter chunkenc.Iterator
 
 	penA, penB int64
 	useA       bool
 }
 
 func newDedupSeriesIterator(a, b adjustableSeriesIterator) *dedupSeriesIterator {
 	return &dedupSeriesIterator{
-		a:     a,
-		b:     b,
-		lastT: math.MinInt64,
-		lastV: float64(math.MinInt64),
-		aval:  a.Next(),
-		bval:  b.Next(),
+		a:        a,
+		b:        b,
+		lastT:    math.MinInt64,
+		lastIter: a,
+		aval:     a.Next(),
+		bval:     b.Next(),
 	}
 }
 
 func (it *dedupSeriesIterator) Next() chunkenc.ValueType {
-	lastValue := it.lastV
+	lastFloatVal, isFloatVal := it.lastFloatVal()
 	lastUseA := it.useA
 	defer func() {
-		if it.useA != lastUseA {
+		if it.useA != lastUseA && isFloatVal {
 			// We switched replicas.
 			// Ensure values are correct bases on value before At.
-			it.adjustAtValue(lastValue)
+			it.adjustAtValue(lastFloatVal)
```

Comment:
I am not sure if we need to adjust histograms here, it would be quite challenging to do it the same way we adjust counters due to dynamic buckets. Let's maybe add a `TODO` and we can investigate later on.